[Inductor] Add Dynamic shape support to user defined triton kernels

oulgen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112801
* #112752
* #112547
* __->__ #112523

1) This PR moves the grid function codegen to wrapper so that we can use
   IndentBuffers as opposed to manually adding tabs for indentation.
2) In inductor, emits the grid function in the body of the kernel call so
   that it can use free symbols from dynamic shapes

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

oulgen committed:
[Inductor] Add Dynamic shape support to user defined triton kernels

1) This PR moves the grid function codegen to wrapper so that we can use
   IndentBuffers as opposed to manually adding tabs for indentation.
2) In inductor, emits the grid function in the body of the kernel call so
   that it can use free symbols from dynamic shapes

[ghstack-poisoned]

oulgen commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -497,14 +514,10 @@ def generate_extern_kernel_out(self, output_view, codegen_reference, args, kerne
-
-            grid, code = grid_fn_code(kernel_name, configs, grid)
-            self.header.splice(code, strip=True)
+        grid, code = user_defined_kernel_grid_fn_code(kernel_name, configs, grid)
+        # Must happen after free symbols are already codegened
+        with self.prefix.indent():
+            self.prefix.splice(code)
```
Looking for feedback here. I moved the grid function to the prefix (rather than the header) so that it can access the free symbols from dynamic shapes. This happens to work because prefix contains the free symbols. Is there a better solution or is there a way for me to assert that free symbols are already generated?

aakhundov replied:
Because the grid function is a function (and not some inlined block of code) would all the symbols defined in the outer scope, including the ones defined *after* this function, not be visible in the inner scope of this function's body? E.g., this works:

```python
def fn(a):
  return a * b

b = 123
c = fn(456)
```

The main thing is that `fn` is *called before* the `b` is defined. But otherwise it's fine that `b` is defined after `fn`.

oulgen replied:
I didn't know this is how python semantics worked.. C++ certainly does not work this way

aakhundov replied:
Depending on how we decide to deal with codegening the grid in the AOTInductor's C++ wrapper codegen, maybe we could opt for relying on this feature of Python.

aakhundov commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -124,6 +124,23 @@ def get_cpp_op_schema(kernel):
+            for grid, c in zip(grids, configs):
+                guards = [f"meta['{name}'] == {val}" for name, val in c.kwargs.items()]
+                guards = " and ".join(guards)
+                output.writeline(f"if {guards}: return {grid}")
```
Should we also generate an exception at the end saying smth like `f"no config found for the {meta=}"`?

oulgen replied:
Triton will error saying that no suitable grid was found for {name}. But yes, we will not know what meta looked like. I could add the exception but by construction aren't we always guaranteed to match? Like what's scenario that we would fail to match?

aakhundov replied:
Not sure about the scenario, more like a defensive / informative code against the unexpected :)

Chillee commented:
Can you provide an example of what the generated code looks like?

oulgen commented:
@Chillee
```
def call(args):
    arg0_1, arg1_1, arg2_1, arg3_1 = args
    args.clear()
    s0 = arg0_1
    assert_size_stride(arg1_1, (s0, ), (1, ))
    assert_size_stride(arg2_1, (s0, ), (1, ))
    def grid_wrapper_for_add_kernel_autotuned_0(meta):
        if meta['BLOCK_SIZE'] == 128: return (((s0 + 127)//128), 1, 1)
        if meta['BLOCK_SIZE'] == 64: return (((s0 + 63)//64), 1, 1)
    with torch.cuda._DeviceGuard(0):
        torch.cuda.set_device(0) # no-op to ensure context
        buf0 = empty((s0, ), device='cuda', dtype=torch.float32)
        # Source Nodes: [output], Original ATen: [aten.zeros_like]
        stream0 = get_cuda_stream(0)
        triton_poi_fused_zeros_like_0.run(buf0, s0, grid=grid(s0), stream=stream0)
        # Source Nodes: [triton_kernel_wrapper_mutation], Original ATen: []
        add_kernel_autotuned_0.run(in_ptr0=arg1_1, in_ptr1=arg2_1, out_ptr=buf0, n_elements=s0, grid=grid_wrapper_for_add_kernel_autotuned_0, stream=stream0)
        del arg1_1
        del arg2_1
        return (buf0, s0, )
```

Full code: P870612754

zou3519 commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -76,25 +76,18 @@ def __init__(self):
-
 @triton_kernel_wrapper_mutation.py_impl(DispatchKey.CompositeExplicitAutograd)
 def triton_kernel_wrapper_mutation_dense(*, kernel_idx, grid, kwargs):
+    from torch._inductor.codegen.wrapper import user_defined_kernel_grid_fn_code
```
I missed some PRs: previously, we were evaluating the grid at Dynamo time. Did that change?

oulgen replied:
We are still doing the same, this is just for emitting the python code for multi option grid function.

From above
> This PR moves the grid function codegen to wrapper so that we can use IndentBuffers as opposed to manually adding tabs for indentation.

zou3519 replied:
gotcha, thanks for clarifying

Chillee commented:
@oulgen And what's the user code for that generated inductor code?

oulgen commented:
@Chillee 
```
        def call_triton(x: torch.Tensor, y: torch.Tensor):
            output = torch.zeros_like(x, requires_grad=grad)
            n_elements = output.numel()
            grid = lambda meta: (triton.cdiv(n_elements, meta["BLOCK_SIZE"]),)
            add_kernel_autotuned[grid](x, y, output, n_elements)
            return output

        t1 = torch.rand(256, device="cuda", requires_grad=grad)
        t2 = torch.rand(256, device="cuda", requires_grad=grad)
```
from the test cases

Is there anything particular you're looking at?

oulgen committed:
Update on "[Inductor] Add Dynamic shape support to user defined triton kernels"


1) This PR moves the grid function codegen to wrapper so that we can use
   IndentBuffers as opposed to manually adding tabs for indentation.
2) In inductor, emits the grid function in the body of the kernel call so
   that it can use free symbols from dynamic shapes

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Chillee reviewed (commented):
How is the grid function evaluated at dynamo time for all of the config options? is that code in this PR?

oulgen commented:
@Chillee 

https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/functions.py#L715-L726

Chillee commented on test/dynamo/test_functions.py:
```diff
@@ -1839,13 +1840,13 @@ def call_triton_return_view(x: torch.Tensor):
         t_view = t.view(16)
 
         compiled_func = torch.compile(
```
Should add some more concrete tests about recompilation here with dynamic shapes.

oulgen replied:
will do as follow up

oulgen commented:
@pytorchbot merge
